### ISSUE TITLE
EMPTY TOKENS preventive check design (NEW TOKEN handling)

### ISSUE THREAD

**Issue by smartguy1196** [smartguy1196/regex.objectified#26]
<strike>As [smartguy1196](https://github.com/smartguy1196) commented [[1]](https://github.com/smartguy1196/regex.objectified/issues/26#issuecomment-244190878), **Issue #24 is a** ***bug,*** **so that needs to be fixed ASAP.** (this will remove the _bug_ label)</strike> - **FIXED!!!**

<strike>He also commented [[2]](https://github.com/smartguy1196/regex.objectified/issues/26#issuecomment-244193572) that **the ESCAPE DETECTOR LOOP contains preventive checks that are _unnecessary_**. This is due to the fact, that right before the SPECIAL ESCAPE DETECTOR condition passes, there is a condition that checks if there is any string at all that follows the ESCAPE character ("\"). **This also needs to be fixed ASAP.** (this will remove the _performance_ label),</strike> - **wontfix**

however, the check that exists before the SPECIAL ESCAPE DETECTOR _only_ checks if there IS NOT a following string. meaning that if there is a string following the ESCAPE CHARACTER ("\"), it could be JUST a special escape sequence, so therefore the _performance_ label is to be removed

This is a significant part of #22 

BEFORE CREATING A NEW TOKEN (using that +1-in-index design in the assignment),

check to be sure that a NEW TOKEN is needed 

AFTER CREATING A NEW TOKEN, and if the check is passed,

increment the STRING TRACKER VALUE to be that of the NEW TOKEN's index

OFFICIAL MODEL IS NOW RELEASED!!!

```
//PREVENTIVE CHECK for NO EMPTY TOKENS creation (if the string after the match is empty, do not create a token for it)
//creates unidentified string tokens for the string following the REGULAR ESCAPE SEQUENCE, if PREVENTIVE CHECK passes

if( this.tokens[ loop.dt_i ].regex.substring( 2 /*uses same logic as BACK REFERENCE ^^^*/ ).length > 0 )
{
  this.tokens[ loop.dt_i+1 ].regex = this.tokens[ loop.dt_i ].regex.substring( 2 /*uses same logic as BACK REFERENCE ^^^*/ );
  this.tokens[ loop.dt_i+1 ].type = "string::unidentified";

  //increase THE TOKENS ARRAY INDEX STORED **WITHIN** the string tracker array

  loop.dt_i = debug.string_tracker[ loop.st_i ] = loop.dt_i + 1;

} //PREVENTIVE CHECK DESIGN (issue #26)
```

**Comment by smartguy1196** [comment]
Issues #17, #24, and #26 need to be merged and kept under #26

**Comment by smartguy1196** [comment]
Issue #17 is a bug, so that needs to be fixed ASAP

**Comment by smartguy1196** [comment]
labelling as a bug temporarily while issue #24 is being fixed

**Comment by smartguy1196** [comment]
labeling as a performance issue as the ESCAPE DETECTOR LOOP contains preventive checks that are uneccessary

This is due to the fact, that right before the SPECIAL ESCAPE DETECTOR condition _passes_, there is a condition that checks if there is any string at all that follows the ESCAPE character ("\")

**Comment by smartguy1196** [comment]
the performance label is temporary

**Comment by smartguy1196** [comment]
look at FIRST comment for reasoning as to why the performance label was removed

**Comment by smartguy1196** [comment]
OFFICIAL MODEL IS NOW RELEASED!!!

```
//PREVENTIVE CHECK for NO EMPTY TOKENS creation (if the string after the match is empty, do not create a token for it)
//creates unidentified string tokens for the string following the REGULAR ESCAPE SEQUENCE, if PREVENTIVE CHECK passes

if( this.tokens[ loop.dt_i ].regex.substring( 2 /*uses same logic as BACK REFERENCE ^^^*/ ).length > 0 )
{
  this.tokens[ loop.dt_i+1 ].regex = this.tokens[ loop.dt_i ].regex.substring( 2 /*uses same logic as BACK REFERENCE ^^^*/ );
  this.tokens[ loop.dt_i+1 ].type = "string::unidentified";

  //increase THE TOKENS ARRAY INDEX STORED **WITHIN** the string tracker array

  loop.dt_i = debug.string_tracker[ loop.st_i ] = loop.dt_i + 1;

} //PREVENTIVE CHECK DESIGN (issue #26)
```

**Comment by smartguy1196** [comment]
bugs are fixed, and model is finished. closing issue

**Comment by smartguy1196** [comment]
removing _fixed_ label as bug is no longer a relevancy

**Comment by smartguy1196** [comment]
Included improvement proposed in #32 into primary comment
